// react-twitter-embed/dist/index.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

function _interopDefault (ex) { return (ex && (typeof ex === 'object') && 'default' in ex) ? ex['default'] : ex; }

var React = require('react');
var React__default = _interopDefault(React);
var PropTypes = _interopDefault(require('prop-types'));

function unwrapExports (x) {
	return x && x.__esModule && Object.prototype.hasOwnProperty.call(x, 'default') ? x.default : x;
}

function createCommonjsModule(fn, module) {
	return module = { exports: {} }, fn(module, module.exports), module.exports;
}

var isRequiredIf_1 = createCommonjsModule(function (module, exports) {
Object.defineProperty(exports,'__esModule',{value:!0});var VALIDATOR_ARG_ERROR_MESSAGE='The typeValidator argument must be a function with the signature function(props, propName, componentName).',MESSAGE_ARG_ERROR_MESSAGE='The error message is optional, but must be a string if provided.',propIsRequired=function propIsRequired(a,b,c,d){if('boolean'==typeof a)return a;return 'function'==typeof a?a(b,c,d):!(!0!==!!a)&&!!a},propExists=function propExists(a,b){return Object.hasOwnProperty.call(a,b)},missingPropError=function missingPropError(a,b,c,d){return d?new Error(d):new Error('Required '+a[b]+' `'+b+'`'+(' was not specified in `'+c+'`.'))},guardAgainstInvalidArgTypes=function guardAgainstInvalidArgTypes(a,b){if('function'!=typeof a)throw new TypeError(VALIDATOR_ARG_ERROR_MESSAGE);if(!!b&&'string'!=typeof b)throw new TypeError(MESSAGE_ARG_ERROR_MESSAGE)},isRequiredIf=function isRequiredIf(a,b,c){return guardAgainstInvalidArgTypes(a,c),function(d,e,f){for(var _len=arguments.length,g=Array(3<_len?_len-3:0),_key=3;_key<_len;_key++)g[_key-3]=arguments[_key];return propIsRequired(b,d,e,f)?propExists(d,e)?a.apply(void 0,[d,e,f].concat(g)):missingPropError(d,e,f,c):a.apply(void 0,[d,e,f].concat(g));// Is not required, so just run typeValidator.
}};exports.default=isRequiredIf;


});

var isRequiredIf = unwrapExports(isRequiredIf_1);

var exenv = createCommonjsModule(function (module) {
/*!
  Copyright (c) 2015 Jed Watson.
  Based on code that is Copyright 2013-2015, Facebook, Inc.
  All rights reserved.
*/
/* global define */

(function () {

	var canUseDOM = !!(
		typeof window !== 'undefined' &&
		window.document &&
		window.document.createElement
	);

	var ExecutionEnvironment = {

		canUseDOM: canUseDOM,

		canUseWorkers: typeof Worker !== 'undefined',

		canUseEventListeners:
			canUseDOM && !!(window.addEventListener || window.attachEvent),

		canUseViewport: canUseDOM && !!window.screen

	};

	if (module.exports) {
		module.exports = ExecutionEnvironment;
	} else {
		window.ExecutionEnvironment = ExecutionEnvironment;
	}

}());
});

var twitter_widget_js = 'https://platform.twitter.com/widgets.js';

var classCallCheck = function (instance, Constructor) {
  if (!(instance instanceof Constructor)) {
    throw new TypeError("Cannot call a class as a function");
  }
};

var createClass = function () {
  function defineProperties(target, props) {
    for (var i = 0; i < props.length; i++) {
      var descriptor = props[i];
      descriptor.enumerable = descriptor.enumerable || false;
      descriptor.configurable = true;
      if ("value" in descriptor) descriptor.writable = true;
      Object.defineProperty(target, descriptor.key, descriptor);
    }
  }

  return function (Constructor, protoProps, staticProps) {
    if (protoProps) defineProperties(Constructor.prototype, protoProps);
    if (staticProps) defineProperties(Constructor, staticProps);
    return Constructor;
  };
}();

var inherits = function (subClass, superClass) {
  if (typeof superClass !== "function" && superClass !== null) {
    throw new TypeError("Super expression must either be null or a function, not " + typeof superClass);
  }

  subClass.prototype = Object.create(superClass && superClass.prototype, {
    constructor: {
      value: subClass,
      enumerable: false,
      writable: true,
      configurable: true
    }
  });
  if (superClass) Object.setPrototypeOf ? Object.setPrototypeOf(subClass, superClass) : subClass.__proto__ = superClass;
};

var possibleConstructorReturn = function (self, call) {
  if (!self) {
    throw new ReferenceError("this hasn't been initialised - super() hasn't been called");
  }

  return call && (typeof call === "object" || typeof call === "function") ? call : self;
};

var TwitterTimelineEmbed = function (_Component) {
     inherits(TwitterTimelineEmbed, _Component);

     function TwitterTimelineEmbed() {
          classCallCheck(this, TwitterTimelineEmbed);
          return possibleConstructorReturn(this, (TwitterTimelineEmbed.__proto__ || Object.getPrototypeOf(TwitterTimelineEmbed)).apply(this, arguments));
     }

     createClass(TwitterTimelineEmbed, [{
          key: 'buildChromeOptions',
          value: function buildChromeOptions(options) {
               options.chrome = '';
               if (this.props.noHeader) {
                    options.chrome = options.chrome + ' noheader';
               }

               if (this.props.noFooter) {
                    options.chrome = options.chrome + ' nofooter';
               }

               if (this.props.noBorders) {
                    options.chrome = options.chrome + ' noborders';
               }

               if (this.props.noScrollbar) {
                    options.chrome = options.chrome + ' noscrollbar';
               }

               if (this.props.transparent) {
                    options.chrome = options.chrome + ' transparent';
               }

               return options;
          }
     }, {
          key: 'buildOptions',
          value: function buildOptions() {
               var options = Object.assign({}, this.props.options);
               if (this.props.autoHeight) {
                    options.height = this.refs.embedContainer.parentNode.offsetHeight;
               }

               options = Object.assign({}, options, {
                    theme: this.props.theme,
                    linkColor: this.props.linkColor,
                    borderColor: this.props.borderColor,
                    lang: this.props.lang
               });

               return options;
          }
     }, {
          key: 'renderWidget',
          value: function renderWidget(options) {
               if (!this.isMountCanceled) {
                    window.twttr.widgets.createTimeline({
                         sourceType: this.props.sourceType,
                         screenName: this.props.screenName,
                         userId: this.props.userId,
                         ownerScreenName: this.props.ownerScreenName,
                         slug: this.props.slug,
                         id: this.props.id || this.props.widgetId,
                         url: this.props.url
                    }, this.refs.embedContainer, options);
               }
          }
     }, {
          key: 'componentDidMount',
          value: function componentDidMount() {
               var _this2 = this;

               if (exenv.canUseDOM) {
                    var script = require('scriptjs');
                    script(twitter_widget_js, 'twitter-embed', function () {
                         if (!window.twttr) {
                              console.error('Failure to load window.twttr in TwitterTimelineEmbed, aborting load.');
                              return;
                         }

                         var options = _this2.buildOptions();
                         /** Append chrome options */
                         options = _this2.buildChromeOptions(options);
                         _this2.renderWidget(options);
                    });
               }
          }
     }, {
          key: 'componentWillUnmount',
          value: function componentWillUnmount() {
               this.isMountCanceled = true;
          }
     }, {
          key: 'render',
          value: function render() {
               return React__default.createElement('div', { ref: 'embedContainer' });
          }
     }]);
     return TwitterTimelineEmbed;
}(React.Component);

TwitterTimelineEmbed.propTypes = {
     /**
          * This can be either of profile, likes, list, collection, URL, widget
          */
     sourceType: PropTypes.oneOf(['profile', 'likes', 'list', 'collection', 'URL', 'widget']).isRequired,
     /**
          * username of twitter handle as String
          */
     screenName: isRequiredIf(PropTypes.string, function (props) {
          return !props.hasOwnProperty('userId') && (props.sourceType === 'profile' || props.sourceType === 'likes');
     }),
     /**
          * UserId of twitter handle as number
          */
     userId: isRequiredIf(PropTypes.number, function (props) {
          return !props.hasOwnProperty('screenName') && (props.sourceType === 'profile' || props.sourceType === 'likes');
     }),
     /**
          * To show list, used along with slug
          */
     ownerScreenName: isRequiredIf(PropTypes.string, function (props) {
          return props.sourceType === 'list' && !props.hasOwnProperty('id');
     }),
     /**
          * To show list, used along with ownerScreenName
          */
     slug: isRequiredIf(PropTypes.string, function (props) {
          return props.sourceType === 'list' && !props.hasOwnProperty('id');
     }),
     /**
          * To show list, unique list id
          * Also used with collections, in that case its valid collection id
          */
     id: isRequiredIf(PropTypes.oneOfType([PropTypes.number, PropTypes.string]), function (props) {
          return props.sourceType === 'list' && !props.hasOwnProperty('ownerScreenName') && !props.hasOwnProperty('slug') || props.sourceType === 'collection';
     }),
     /**
          * To show twitter content with url.
          * Supported content includes profiles, likes, lists, and collections.
          */
     url: isRequiredIf(PropTypes.string, function (props) {
          return props.sourceType === 'url';
     }),
     /**
          * To show custom widget
          */
     widgetId: isRequiredIf(PropTypes.string, function (props) {
          return props.sourceType === 'widget';
     }),
     /**
          * Additional options to pass to twitter widget plugin
          */
     options: PropTypes.object,
     /**
          * Automatically fit into parent container height
          */
     autoHeight: PropTypes.bool,
     /**
          * With dark or light theme
          */
     theme: PropTypes.oneOf(['dark', 'light']),
     /**
          * With custom link colors. Note: Only Hex colors are supported.
          */
     linkColor: PropTypes.string,
     /**
          * With custom border colors. Note: Only Hex colors are supported.
          */
     borderColor: PropTypes.string,
     /**
          * Hide the header from timeline
          */
     noHeader: PropTypes.bool,
     /**
          * Hide the footer from timeline
          */
     noFooter: PropTypes.bool,
     /**
          * Hide the border from timeline
          */
     noBorders: PropTypes.bool,
     /**
          * Hide the scrollbars
          */
     noScrollbar: PropTypes.bool,
     /**
          * Enable Transparancy
          */
     transparent: PropTypes.bool,
     /**
          * Custom language code. Supported codes here: https://developer.twitter.com/en/docs/twitter-for-websites/twitter-for-websites-supported-languages/overview.html
          */
     lang: PropTypes.string
};

var TwitterShareButton = function (_Component) {
  inherits(TwitterShareButton, _Component);

  function TwitterShareButton() {
    classCallCheck(this, TwitterShareButton);
    return possibleConstructorReturn(this, (TwitterShareButton.__proto__ || Object.getPrototypeOf(TwitterShareButton)).apply(this, arguments));
  }

  createClass(TwitterShareButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterShareButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createShareButton(_this2.props.url, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterShareButton;
}(React.Component);

TwitterShareButton.propTypes = {
  /**
  * Url for sharing
  */
  url: PropTypes.string.isRequired,
  /**
  * Additional options for overriding config. Details at : https://dev.twitter.com/web/tweet-button/parameters
  */
  options: PropTypes.object
};

var TwitterFollowButton = function (_Component) {
  inherits(TwitterFollowButton, _Component);

  function TwitterFollowButton() {
    classCallCheck(this, TwitterFollowButton);
    return possibleConstructorReturn(this, (TwitterFollowButton.__proto__ || Object.getPrototypeOf(TwitterFollowButton)).apply(this, arguments));
  }

  createClass(TwitterFollowButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterFollowButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createFollowButton(_this2.props.screenName, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterFollowButton;
}(React.Component);

TwitterFollowButton.propTypes = {
  /**
       * Username of twitter user which will be followed on click
       */
  screenName: PropTypes.string.isRequired,
  /**
       * Additional options to be added to the button
       */
  options: PropTypes.object
};

var TwitterHashtagButton = function (_Component) {
  inherits(TwitterHashtagButton, _Component);

  function TwitterHashtagButton() {
    classCallCheck(this, TwitterHashtagButton);
    return possibleConstructorReturn(this, (TwitterHashtagButton.__proto__ || Object.getPrototypeOf(TwitterHashtagButton)).apply(this, arguments));
  }

  createClass(TwitterHashtagButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterHashtagButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createHashtagButton(_this2.props.tag, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterHashtagButton;
}(React.Component);

TwitterHashtagButton.propTypes = {
  /**
       * Tag name for hashtag button
       */
  tag: PropTypes.string.isRequired,
  /**
       * Additional options to be added to the button
       */
  options: PropTypes.object
};

var TwitterMentionButton = function (_Component) {
  inherits(TwitterMentionButton, _Component);

  function TwitterMentionButton() {
    classCallCheck(this, TwitterMentionButton);
    return possibleConstructorReturn(this, (TwitterMentionButton.__proto__ || Object.getPrototypeOf(TwitterMentionButton)).apply(this, arguments));
  }

  createClass(TwitterMentionButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterMentionButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createMentionButton(_this2.props.screenName, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterMentionButton;
}(React.Component);

TwitterMentionButton.propTypes = {
  /**
   * Username to which you will need to tweet
   */
  screenName: PropTypes.string.isRequired,
  /**
   * Additional options for overriding config.
   */
  options: PropTypes.object
};

var TwitterTweetEmbed = function (_Component) {
  inherits(TwitterTweetEmbed, _Component);

  function TwitterTweetEmbed() {
    classCallCheck(this, TwitterTweetEmbed);
    return possibleConstructorReturn(this, (TwitterTweetEmbed.__proto__ || Object.getPrototypeOf(TwitterTweetEmbed)).apply(this, arguments));
  }

  createClass(TwitterTweetEmbed, [{
    key: 'renderWidget',
    value: function renderWidget() {
      if (!window.twttr) {
        console.error('Failure to load window.twttr in TwitterTweetEmbed, aborting load.');
        return;
      }
      if (!this.isMountCanceled) {
        window.twttr.widgets.createTweet(this.props.tweetId, this.refs.embedContainer, this.props.options);
      }
    }
  }, {
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          _this2.renderWidget();
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterTweetEmbed;
}(React.Component);

TwitterTweetEmbed.propTypes = {
  /**
       * Tweet id that needs to be shown
       */
  tweetId: PropTypes.string.isRequired,
  /**
       * Additional options to pass to twitter widget plugin
       */
  options: PropTypes.object
};

var TwitterMomentShare = function (_Component) {
  inherits(TwitterMomentShare, _Component);

  function TwitterMomentShare() {
    classCallCheck(this, TwitterMomentShare);
    return possibleConstructorReturn(this, (TwitterMomentShare.__proto__ || Object.getPrototypeOf(TwitterMomentShare)).apply(this, arguments));
  }

  createClass(TwitterMomentShare, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterMomentShare, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createMoment(_this2.props.momentId, _this2.refs.shareMoment, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'shareMoment' });
    }
  }]);
  return TwitterMomentShare;
}(React.Component);

TwitterMomentShare.propTypes = {
  /**
   * id of Twitter moment to show
   */
  momentId: PropTypes.string.isRequired,
  /**
   * Additional options for overriding config.
   */
  options: PropTypes.object
};

var TwitterDMButton = function (_Component) {
  inherits(TwitterDMButton, _Component);

  function TwitterDMButton() {
    classCallCheck(this, TwitterDMButton);
    return possibleConstructorReturn(this, (TwitterDMButton.__proto__ || Object.getPrototypeOf(TwitterDMButton)).apply(this, arguments));
  }

  createClass(TwitterDMButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterDMButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createDMButton(_this2.props.id, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterDMButton;
}(React.Component);

TwitterDMButton.propTypes = {
  /**
       * Twitter user id for DM button
       */
  id: PropTypes.number.isRequired,
  /**
       * Additional options to be added to the button
       */
  options: PropTypes.object
};

var TwitterVideoEmbed = function (_Component) {
  inherits(TwitterVideoEmbed, _Component);

  function TwitterVideoEmbed() {
    classCallCheck(this, TwitterVideoEmbed);
    return possibleConstructorReturn(this, (TwitterVideoEmbed.__proto__ || Object.getPrototypeOf(TwitterVideoEmbed)).apply(this, arguments));
  }

  createClass(TwitterVideoEmbed, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterVideoEmbed, aborting load.');
            return;
          }
          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createVideo(_this2.props.id, _this2.refs.embedContainer);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterVideoEmbed;
}(React.Component);

TwitterVideoEmbed.propTypes = {
  /**
       * Id of video tweet.
       */
  id: PropTypes.string.isRequired
};

var TwitterOnAirButton = function (_Component) {
  inherits(TwitterOnAirButton, _Component);

  function TwitterOnAirButton() {
    classCallCheck(this, TwitterOnAirButton);
    return possibleConstructorReturn(this, (TwitterOnAirButton.__proto__ || Object.getPrototypeOf(TwitterOnAirButton)).apply(this, arguments));
  }

  createClass(TwitterOnAirButton, [{
    key: 'componentDidMount',
    value: function componentDidMount() {
      var _this2 = this;

      if (exenv.canUseDOM) {
        var script = require('scriptjs');
        script(twitter_widget_js, 'twitter-embed', function () {
          if (!window.twttr) {
            console.error('Failure to load window.twttr in TwitterOnAirButton, aborting load.');
            return;
          }

          if (!_this2.isMountCanceled) {
            window.twttr.widgets.createPeriscopeOnAirButton(_this2.props.username, _this2.refs.embedContainer, _this2.props.options);
          }
        });
      }
    }
  }, {
    key: 'componentWillUnmount',
    value: function componentWillUnmount() {
      this.isMountCanceled = true;
    }
  }, {
    key: 'render',
    value: function render() {
      return React__default.createElement('div', { ref: 'embedContainer' });
    }
  }]);
  return TwitterOnAirButton;
}(React.Component);

TwitterOnAirButton.propTypes = {
  /**
   * Username for which you require periscope on air button
   */
  username: PropTypes.string.isRequired,
  /**
   * Additional options for overriding config.
   */
  options: PropTypes.object
};

exports.TwitterTimelineEmbed = TwitterTimelineEmbed;
exports.TwitterShareButton = TwitterShareButton;
exports.TwitterFollowButton = TwitterFollowButton;
exports.TwitterHashtagButton = TwitterHashtagButton;
exports.TwitterMentionButton = TwitterMentionButton;
exports.TwitterTweetEmbed = TwitterTweetEmbed;
exports.TwitterMomentShare = TwitterMomentShare;
exports.TwitterDMButton = TwitterDMButton;
exports.TwitterVideoEmbed = TwitterVideoEmbed;
exports.TwitterOnAirButton = TwitterOnAirButton;
